### ISSUE TITLE
Unicode characters mangled in RRset comments retrieved through the API

### ISSUE THREAD

**Issue by uedvt359** [PowerDNS/pdns#13050]
- Program: Authoritative
 - Issue type: Bug report

### Short description
RRset comments containing Unicode (UTF-8) characters are mangled on retrieval. They are correctly stored in the database.

### Environment

 - Operating system: RHEL 8.8
 - Software version: pdns-4.7.4-1pdns.el8.x86_64
 - Software source: https://repo.powerdns.com/repo-files/el-auth-47.repo

### Steps to reproduce
1. Insert a RRset comment, e.g. by using this API call:
```
curl -X PATCH --data '{"rrsets": [ {"comments": [ { "account": "uedvt359", "content": "Motörhead" } ], "name": "motorhead.example.org.", "type": "A", "changetype": "REPLACE", "records": [ {"content": "8.8.8.8", "disabled": false } ], "ttl": 900 } ] }' -H 'X-API-Key: ...' http://localhost:8081/api/v1/servers/localhost/zones/example.org.
```
2. verify that the comment ended up in the Database correctly:
```
# psql -Upostgres -dpowerdns
powerdns=# select * from comments where name like 'motorhead.%';
  id   | domain_id |         name          | type | modified_at | account  | comment  
-------+-----------+-----------------------+------+-------------+----------+----------
 81337 |       564 | motorhead.example.org | A    |  1689757634 | uedvt359 | Motörhead
```
3. retrive the RRset again from the API
```
curl http://localhost:8081/api/v1/servers/localhost/zones/example.org. -H "X-API-Key: ..."|jq .
// ...
  "rrsets": [
    {
      "comments": [
        {
          "account": "uedvt359",
          "content": "Mot￿f￿frhead",
          "modified_at": 1689758495
        }
      ],
// ...
```

### Expected behaviour
The comment says `Motörhead`.

### Actual behaviour
The comment says `Mot￿f￿frhead`. The `ö` is replaced by `\u{ffff}f\u{ffff}f` (the invalid noncharacter, followed by a literal lowercase f; twice).

**Comment by zeha** [comment]
Which backend and database are used? Which character set settings are in use in those?

**Comment by uedvt359** [comment]
Backend is gpgsql; database is postgresql-10.23 with UTF8 encoding.

**Comment by rgacogne** [comment]
Right, we tried to make the situation better in https://github.com/PowerDNS/pdns/pull/12260 but this was perhaps not enough.

**Comment by uedvt359** [comment]
Just stating the obvious, but it seems that there are two issues here: one is that the json encoder doesn't handle characters from the supplementary planes, the other is that non-ascii characters get replaced by 0x0fffff.

I've captured the packets between postgres and pdns. the data is transmitted between the two as proper UTF-8:

```
    0000104D  33 35 39 00 00 00 0a 4d  6f 74 c3 b6 72 68 65 61   359....M ot..rhea
    0000105D  64 44 00 00 00 64 00 06  00 00 00 03 35 36 34 00   dD...d.. ....564.
```

**Comment by uedvt359** [comment]
I had a look at the relevant code: https://github.com/PowerDNS/pdns/blob/53c96ca8fe2126e8a0782375db0210377e5be9fc/ext/json11/json11.cpp#L96-L102

Ignoring the issue of non-ascii characters getting turned into 0x0fffff for now, the json encoder is completely broken. It mangles any bytes > 0x7f (i.e. multi byte utf8 sequences). The patch you linked says this was done to 'escape' binary data in JSON strings. JSON cannot handle binary strings, it only supports well-formed Unicode text. Treating bytes as codepoints, when the input is utf8, breaks all non-ascii characters (e.g. `ö` in utf8 is the two byte sequence `c3 b6`. 'escaping' this as `\u00c3\u00b6` turns this into two seperate code points `Ã¶`). Going back to #9349 and #12260, the proper fix would have been to discard invalid http requests (uris may only contain a subset of printable ascii characters).

Regarding the other problem I've done some tests and one `\ufffff` sequence is generated for each byte of a utf-8 character (`ö` two bytes, `€` three, `𠜎` four). the reason for this rather odd private-use character is probably the `char buf[8]`: `snprintf()` may write at most 7 characters plus the terminating null byte into this buffer. `%x` expects an unsigned int (i.e. `uint32_t`), but is passed a signed char (`const char ch = value[i]`; `int8_t`) instead that happens to be interpreted as negative on multi byte characters (as they have their high bits set). The char `ch` is then sign-extended from `0xc3` to `0xffffffc3`; and then read from snprintf, who only has space to print `fffff` before it runs out of buffer space.

In short, the following needs to be done to fix this:
1. revert #12260' changes to ext/json11/json11.cpp
2. in pdns/webserver.cc `WebServer::handleRequest()` (or possibly even earlier in `WebServer::serveConnection`) should drop any requests with malformed uris
3. notify the `yahttp` authors that their URL parser accepts invalid urls (https://github.com/cmouse/yahttp/blob/master/yahttp/url.hpp#L79).

**Comment by omoerbeek** [comment]
Thank yo for your analysis. I have to admit that I 12260 is not a proper fix and maybe made things worse.
When working on this, we do have to solve the issue that e.g. the statistics endpoint can produce arbitrary data, not only lines with malformed URIs.
